lots of opinionated changes!

rogpeppe opened the pull request:
I like the general approach, but I had a bunch of ideas about how
things might be improved. This PR lumps 'em all together...

The main API changes are:
-  lose the default client - unlike with http.Client, I don't think a global client
    is the right thing here, because we are addressing one particular server
    and producing stats designed for that server. In preference, I'd choose to
    create domain-specific statsd.Client global variables where desired
    outside of this package, making it possible for different components to
    generate independent stats to different servers.
- connect at initialisation time rather than on first write - "connecting" on a UDP socket
    doesn't involve actually contacting the remote server (just a potential DNS lookup) so
    I think it makes sense to do this earlier rather than later, as this is the most likely
    time that we might see an error. This simplifies the code logic
    too.
- don't return errors from all methods. Since this is designed to be used as a global variable
    and writes are amortised, there is unlikely to be anything a client can usefully do if
    a stat call returns an error. Instead, we provide an error callback that can be used to
    inform of errors on writes. In practice, this will probably never be invoked as UDP will
    silently drop packets rather than return an error.

Other more minor API changes:
- dropped the Timing method - it seemed redundant with Duration, and a touch confusing.
- lose the rate arguments from Unique and Gauge, as the server (the etsy statsd at least)
  ignores them.

I also made various changes to the implementation in the interests of efficiency (we
want to be able to generate stats with very low overhead). With a nop writer, the new code
is about 80% faster than the old code (272ns vs 1398ns) and does not allocate at all (the old code incurs
at least 4 allocs per update).

Even if you think it seems reasonable, this PR is not yet ready for submission (I'd like a few more tests to bring test coverage up to 100%, and the docs need work) but I thought this is probably a good place to stop and ask for thoughts and reactions.

Further work:
- I'd really like to see some tests that work against a live statsd server so that we have some idea of whether this code will actually work in reality.
- I think there should be a flusher that periodically flushes stats after a timeout, so that we don't have stats sitting in the buffer and getting indefinitely stale.

cmars commented:
Lots of good stuff here, I really like where its going! Looking forward to landing many of these improvements.

> -  lose the default client - unlike with http.Client, I don't think a global client
    is the right thing here, because we are addressing one particular server
    and producing stats designed for that server. In preference, I'd choose to
    create domain-specific statsd.Client global variables where desired
    outside of this package, making it possible for different components to
    generate independent stats to different servers.

Good point.. more freedom for the package importer there.

> - connect at initialisation time rather than on first write - "connecting" on a UDP socket
    doesn't involve actually contacting the remote server (just a potential DNS lookup) so
    I think it makes sense to do this earlier rather than later, as this is the most likely
    time that we might see an error. This simplifies the code logic
    too.

Not sure I agree with this one, but could use some advice. I've noticed in Go that if there is nothing listening at the configured UDP address when the client does net.Dial, the client connection is never able to write to this endpoint, even if it later shows up. Shouldn't be a problem since UDP is stateless, but it is for Go's UDP implementation. This was detected and fixed with my `TestReconnect` test.

This is why there's a "reconnect" on write error in log/syslog[1]. I think we'll need something similar. Our client needs to be resilient to server outages, including at startup.

[1] http://golang.org/src/log/syslog/syslog.go?s=3915:3960#L251

> I also made various changes to the implementation in the interests of efficiency (we
want to be able to generate stats with very low overhead). With a nop writer, the new code
is about 80% faster than the old code (272ns vs 1398ns) and does not allocate at all (the old code incurs
at least 4 allocs per update).

Amazing! Thanks for this!

> Further work:
- I'd really like to see some tests that work against a live statsd server so that we have some idea of whether this code will actually work in reality.

I decided to stand up a really simple one in my httpstats PR #1. We might want a "real" one for integration tests though :)

> - I think there should be a flusher that periodically flushes stats after a timeout, so that we don't have stats sitting in the buffer and getting indefinitely stale.

+100

rogpeppe commented:
"Not sure I agree with this one, but could use some advice. I've noticed in Go that if there is nothing listening at the configured UDP address when the client does net.Dial, the client connection is never able to write to this endpoint, even if it later shows up."

That would seem very strange behaviour for a UDP connection, as the UDP "connection" doesn't involve the exchange of any packets AFAIK. I can't duplicate the behaviour - this code works as expected for me:
http://play.golang.org/p/bULuZURuPS

cmars commented:
I'm not able to reproduce the behavior I thought I was seeing, using your simple example. Perhaps dial-up-front is fine..

But I'm still concerned about why the log/syslog implementation would reconnect on a UDP write error (see log/syslog/Writer.writeAndRetry).. Why would it do that? Seems like they're anticipating a case where the UDPConn goes bad.

rogpeppe commented:
AFAICS the log/syslog implementation is network independent,
so it can work with TCP-based implementations too, where
it would be necessary to redial.

cmars commented:
Ok, that makes sense. Thanks for the clarification!

rogpeppe commented:
I've updated this PR to improve the documentation and add full test coverage.
I also changed the name of SetAddr to SetHostPort so the nature of its string argument
is hopefully clearer.

PTAL.

cmars commented on statsd.go:
```diff
@@ -29,156 +32,218 @@ const (
 )
 
-type client struct {
+type Client struct {
```
Please add a godoc comment.

cmars commented:
LGTM with a trivial godoc. Hard to disagree with the "opinion" of:
- optional error handling
- autocorrection of absolute gauge values
- performance & benchmarks
- 100% test coverage

Just amazing. Thanks!
